snapcraft/internal/repo/_deb.py

```diff
@@ -252,26 +252,52 @@ def get_installed_build_packages(cls, package_names):
                     build_packages.pop(0))
                 if package_name.endswith(':any'):
                     package_name = package_name[:-4]
-                try:
-                    installed_package = apt_cache[package_name].candidate
-                except KeyError as e:
-                    providers = apt_cache.get_providing_packages(package_name)
-                    if providers:
-                        installed_package = providers[0].candidate
-                    else:
-                        raise errors.BuildPackageNotFoundError(e) from e
+                installed_package = cls._get_providing_package(
+                    apt_cache, package_name)
                 if str(installed_package) not in installed_packages:
                     installed_packages.append(str(installed_package))
-                    for depends in installed_package.get_dependencies(
-                            'Depends'):
-                        # deps is a list of or dependencies. We are taking
-                        # the first one that satisfies the dependency, which
-                        # might or might not be problematic.
-                        # --elopio - 20170504
-                        build_packages.append(depends[0].name)
+                    dependencies = cls._get_dependencies(
+                        apt_cache, installed_package)
+                    build_packages.extend(
+                        [dependency for dependency in dependencies
+                         if dependency not in installed_packages])
+
         return installed_packages
 
     @classmethod
+    def _get_providing_package(cls, apt_cache, package_name):
+        try:
+            return apt_cache[package_name].candidate
+        except KeyError as e:
+            # If the package is not in the cache, it might be a virtual package
+            # provided by another one.
+            providers = apt_cache.get_providing_packages(package_name)
+            if providers:
+                return providers[0].candidate
+            else:
+                raise errors.BuildPackageNotFoundError(e) from e
+
+    @classmethod
+    def _get_dependencies(cls, apt_cache, package):
+        dependencies = []
+        for depends in package.get_dependencies(
+                'Depends'):
+            # Depends is a list of or dependencies. Some might not
+            # be in the cache, so search for the first one in the
+            # cache.
+            for depend in depends:
+                try:
+                    dependency = cls._get_providing_package(
+                        apt_cache, depend.name)
+                    dependencies.append(dependency.package.name)
+                    break
+                except errors.BuildPackageNotFoundError:
+                    pass
+            else:
+                raise errors.BuildPackageNotFoundError(depends)
```

Comment:
This looks a little ugly to me for two reasons:

1. You're looping through the list, breaking the loop **unless** there's an exception in which case... the loop continues. This is awkward.
    Can we check if the package is in the cache instead of the _try/ except_?
2. You're raising **BuildPackageNotFoundError** with a list of packages - the exception takes a single package name.
    Perhaps we need a new exception, like BuildDependenciesMissingError, that takes a list. In the future we could also try harder to list all missing packages rather than just the first one.